### PR TITLE
chore: refactor pages dev to use unstable_dev

### LINKED ISSUE #70: WebSocket support for wrangler pages dev
WebSockets aren't working on pages dev at the moment. Probably something to do with the way we're MITM-ing requests.

### LINKED ISSUE #1347: 🚀 Feature Request: Support selecting HTTP/HTTPS for pages dev
### Describe the solution

We should be support starting local dev (`wrangler pages dev`) with HTTPS. Wrangler local dev (`wrangler dev --local`) already supports this through `--local-protocol` (and `local_protocol` in wrangler.toml).

### PR DESCRIPTION
**Pull request by Skye-31**
Refactor `wrangler pages dev` to use Wrangler-proper's own dev server.

This:

- fixes some bugs (e.g. not proxying WebSockets correctly) (fixes #70),
- presents a much nicer UI (with the slick keybinding controls),
- adds features that `pages dev` was missing (e.g. `--local-protocol`) (fixes #1347),
- and reduces the maintenance burden of `wrangler pages dev` going forward.

### CODE REVIEW

File: packages/wrangler/src/miniflare-cli/assets.ts [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+	const getAsset = (path: string) => {
+		if (assetExists(path)) {
+			return join(directory, path);
+		}
+	};
+
+	let redirectsMatcher = generateRedirectsMatcher(redirectsFile);
+	let headersMatcher = generateHeadersMatcher(headersFile);
+
+	// watch([headersFile, redirectsFile], {
```
Comment (Skye-31) [inline packages/wrangler/src/miniflare-cli/assets.ts:315]: this needs to be looked at

**Comment by threepointone** [comment]
chatted with @GregBrimble, marking as a draft

**Comment by GregBrimble** [comment]
Spoke to @rozenmd who gave a conceptual thumbs-up to this. We spoke about moving `logLevel` and `disableLogs` into the new `miniflareCLIOptions`, but having now attempted that, I realize it's being used my the wrangler process as well, so tidying it away into `miniflareCLIOptions` actually just muddies the waters. I've left it as-is.

@threepointone, this is now ready for review.

File: packages/wrangler/scripts/bundle.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ async function buildMiniflareCLI(flags: BuildFlags = {}) {
 		outfile: "./miniflare-dist/index.mjs",
 		platform: "node",
 		format: "esm",
-		external: ["miniflare", "@miniflare/core"],
+		external: ["miniflare", "@miniflare/core", "chokidar"],
```
Comment (threepointone) [inline packages/wrangler/scripts/bundle.ts:51]: could you add this to scripts/deps.ts to `EXTERNAL_DEPENDENCIES` as well? Honestly, maybe we should just use that array here

File: packages/wrangler/src/dev.tsx [hunk 3]
```diff
@@ -381,34 +416,47 @@ export async function startDev(args: ArgumentsCamelCase<DevArgs>) {
 		async function getBindings(
 			configParam: Config
 		): Promise<CfWorkerInit["bindings"]> {
+			configParam.kv_namespaces ??= [];
+			configParam.durable_objects ??= { bindings: [] };
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:420]: why this? we don't want to mutate the config objects I think.

File: packages/wrangler/src/dev.tsx [hunk 4]
```diff
@@ -505,9 +554,13 @@ export async function startDev(args: ArgumentsCamelCase<DevArgs>) {
 					bindings={bindings}
 					crons={config.triggers.crons}
 					logLevel={args.logLevel}
+					logPrefix={args.logPrefix}
 					onReady={args.onReady}
 					inspect={args.inspect ?? true}
 					showInteractiveDevSession={args.showInteractiveDevSession}
+					forceLocal={args.forceLocal}
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:561]: what is forceLocal, and why can't we use initialMode here?

File: packages/wrangler/src/dev.tsx [hunk 5]
```diff
@@ -505,9 +554,13 @@ export async function startDev(args: ArgumentsCamelCase<DevArgs>) {
...
+					cfFetch={args.cfFetch}
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:562]: what is cfFetch?

File: packages/wrangler/src/api/dev.ts [hunk 6]
```diff
@@ -13,39 +14,73 @@ interface DevOptions {
...
+		script_name?: string | undefined;
+		environment?: string | undefined;
+	}[];
+	showInteractiveDevSession?: boolean;
+	logLevel?: "none" | "error" | "log" | "warn" | "debug";
+	logPrefix?: string;
+	inspect?: boolean;
+	forceLocal?: boolean;
+	cfFetch?: boolean;
+	miniflareCLIOptions?: MiniflareCLIOptions;
```
Comment (threepointone) [inline packages/wrangler/src/api/dev.ts:41]: I think this shouldn't be so generic named, let's just use enableAssetsServiceBinding as the prop to pass ( and it should probably be called `enablePagesAssetsServiceBinding`

File: packages/wrangler/src/api/dev.ts [hunk 7]
```diff
@@ -13,39 +14,73 @@ interface DevOptions {
...
  *  @param {DevOptions} options
  */
-export async function unstable_dev(script: string, options: DevOptions) {
-	logger.warn(
-		`unstable_dev() is experimental\nunstable_dev()'s behaviour will likely change in future releases`
-	);
+export async function unstable_dev(
+	script: string,
+	options: DevOptions,
+	disableWarning?: boolean
```
Comment (threepointone) [inline packages/wrangler/src/api/dev.ts:53]: Let's rename this to disableExperimentaWarning

File: packages/wrangler/templates/pages-shim.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export default {
```
Comment (threepointone) [inline packages/wrangler/templates/pages-shim.ts:1]: Add a comment here saying what this file does? 
(Also note to self, do the same for the other templates)

**Review by threepointone — COMMENTED** [review]
This is looking good, but I had some questions. I'll dive in further once I understand further!

File: packages/wrangler/src/dev.tsx (hunk 4 above)
Comment (GregBrimble) [reply packages/wrangler/src/dev.tsx:561]: `forceLocal` prevents users from ever switching to remote mode. It disables the hotkey and hides it from the UI.

File: packages/wrangler/src/api/dev.ts (hunk 6 above)
Comment (GregBrimble) [reply packages/wrangler/src/api/dev.ts:41]: Yup, happy to change this if you prefer it. I had specifically tried to make it something more general that could be re-used, as I figured there's probably only going to be more things like this, but can easily make it a Pages-specific thing right now.

File: packages/wrangler/src/dev.tsx (hunk 3 above)
Comment (GregBrimble) [reply packages/wrangler/src/dev.tsx:420]: Yeah, bit of a shortcut. I can rewrite that to be cleaner.

**Comment by threepointone** [comment]
I didn't realise we're blocking lands on code coverage stats. @JacobMGEvans could we disable that please?

**Comment by JacobMGEvans** [comment]
> I didn't realise we're blocking lands on code coverage stats. @JacobMGEvans could we disable that please?

Oh it shouldn't be blocking, I'll take a look at reconfigure it.

File: packages/wrangler/src/dev.tsx [hunk 9]
```diff
@@ -250,7 +259,29 @@ export async function devHandler(args: ArgumentsCamelCase<DevArgs>) {
...
+		preview_id?: string;
+	}[];
+	durableObjects?: {
+		name: string;
+		class_name: string;
+		script_name?: string | undefined;
+		environment?: string | undefined;
+	}[];
+	forceLocal?: boolean;
+	cfFetch?: boolean;
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:280]:
```suggestion
```

**Review by threepointone — APPROVED** [review]

**Pull request merged**